### PR TITLE
[zeroc-ice] New port

### LINKED ISSUE #6106: ZeroC Ice Support
Is there any interest or possibility in making a port for this?

https://github.com/zeroc-ice/ice

Thanks in advance.

### PR DESCRIPTION
**Pull request by bold84**
**Describe the pull request**

- #### What does your PR fix?
  Fixes #6106

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  windows, yes (still working on linux / mac)

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  `Yes`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by bold84** [comment]
The build succeeds, but this is an issue I do not know how to resolve.

Ice requires the usage of an IDL file compiler, which I download in the powershell script before the build starts. 

vcpkg_install_msbuild() doesn't let me remove these files before "installing" the binaries to the package folder. 
I could use vcpkg_execute_build_process() instead of vcpkg_install_msbuild() but I'm not sure if that's okay for you guys.


Screen Shot 2022-03-25 at 14 21 08

**Comment by LilyWangLL** [comment]
Now the pipeline error is `The folder /include is empty or not present. This indicates the library was not correctly installed.` the header file install incorrect.

**Comment by bold84** [comment]
> Now the pipeline error is `The folder /include is empty or not present. This indicates the library was not correctly installed.` the header file install incorrect.

Fixed...

**Comment by LilyWangLL** [comment]
`zeroc-ice:x64-windows-static` build failed on pipeline with the following error:
```
-- Installing: D:/packages/zeroc-ice_x64-windows-static/share/zeroc-ice/copyright
-- Performing post-build validation
Invalid crt linkage. Expected Debug,Static, but the following libs had:

    D:\packages\zeroc-ice_x64-windows-static\debug\lib\iceutild.lib: Debug,Dynamic
    D:\packages\zeroc-ice_x64-windows-static\debug\lib\sliced.lib: Debug,Dynamic

To inspect the lib files, use:
    dumpbin.exe /directives mylibfile.lib
Invalid crt linkage. Expected Release,Static, but the following libs had:

    D:\packages\zeroc-ice_x64-windows-static\lib\iceutil.lib: Release,Dynamic
    D:\packages\zeroc-ice_x64-windows-static\lib\slice.lib: Release,Dynamic

To inspect the lib files, use:
    dumpbin.exe /directives mylibfile.lib
Found 2 post-build check problem(s). To submit these ports to curated catalogs, please first correct the portfile:
    C:\a\1\s\ports\zeroc-ice\portfile.cmake
```

**Comment by ZeroAbility** [comment]
If I recall correctly, the CRT linkage was the reason why I opted to integrate CMake into Ice. There are also slice files that need to be available for the FindIce() feature to work correctly and I don't recall if the MSBuild recipe worked with that. I recall the cleanup in vcpkg will delete some binary files.

**Comment by bold84** [comment]
> If I recall correctly, the CRT linkage was the reason why I opted to integrate CMake into Ice. There are also slice files that need to be available for the FindIce() feature to work correctly and I don't recall if the MSBuild recipe worked with that. I recall the cleanup in vcpkg will delete some binary files.

I removed the hard coded crt linkage from the project files. 

But good point, I need to add the slice files to the installation procedure. But I first have to get Linux and Mac working properly.

**Comment by ZeroAbility** [comment]
I hope this will help you:

https://github.com/mumble-voip/mumble/tree/master/helpers/vcpkg/ports/zeroc-ice

You may still need to set `-DIce_Home` for FindIce() to work because of how it looks for it in Windows. 

Here is the CMake integrated fork of Ice if you want to use some of the info there and can recreate what is needed with MSBuild:

https://github.com/mumble-voip/ice

**Comment by bold84** [comment]
Thank you! :-)
I've seen this before already, but I preferred to stick with what ZeroC provides instead of using your CMake scripts.

**Comment by ZeroAbility** [comment]
> Thank you! :-) I've seen this before already, but I preferred to stick with what ZeroC provides instead of using your CMake scripts.

You're welcome. Good luck!

**Review by LilyWangLL — APPROVED** [review]

File: ports/zeroc-ice/vcpkg.json [hunk 1]
```diff
@@ -0,0 +1,185 @@
+{
+  "name": "zeroc-ice",
+  "version": "3.7.7",
+  "description": "Comprehensive RPC framework with support for C++, CSharp, Java, JavaScript, Python and more.",
+  "homepage": "https://github.com/zeroc-ice/ice",
+  "license": null,
```
Comment (Krzmbrzl) [inline ports/zeroc-ice/vcpkg.json:6]: Shouldn't this say GPL according to https://github.com/zeroc-ice/ice/blob/3.7/LICENSE?

File: scripts/ci.baseline.txt [hunk 2]
```diff
@@ -1439,3 +1439,11 @@ workflow:arm-uwp=fail
...
 dimcli:x64-windows-static-md=fail
 dimcli:x64-windows-static=fail
+
+# ZeroC doesn't provide ARM tagets in project files.
+zeroc-ice:arm64-windows=fail
+zeroc-ice:arm-uwp=fail
+zeroc-ice:x64-uwp=fail
+zeroc-ice:x64-windows-static=fail
+zeroc-ice:x64-linux=fail
+zeroc-ice:x64-osx=fail
```
Comment (Krzmbrzl) [inline scripts/ci.baseline.txt:1449]: Doesn't this port work in these cases? The cmake scripts at least seam to cover non-Windows cases and also the static case :eyes:

File: ports/zeroc-ice/slice2swift.patch [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ {
+@@ -1531,11 +1531,6 @@ SwiftGenerator::writeMemberwiseInitializer(IceUtilInternal::Output& out,
+     {
+         out << sp;
+         out << nl;
+-        int typeCtx = TypeContextInParam;
+-        if(local)
+-        {
+-            typeCtx |= TypeContextLocal;
+-        }
```
Comment (strega-nil-ms) [inline ports/zeroc-ice/slice2swift.patch:22]: why is this necessary? this seems to break an important feature...

File: ports/zeroc-ice/md5i_fix.patch [hunk 4]
```diff
@@ -0,0 +1,13 @@
+diff --git a/cpp/src/Slice/MD5I.cpp b/cpp/src/Slice/MD5I.cpp
+index 66f4de545b..2430b1fd1b 100644
+--- a/cpp/src/Slice/MD5I.cpp
++++ b/cpp/src/Slice/MD5I.cpp
+@@ -172,7 +172,7 @@ md5_process(md5_state_t *pms, const md5_byte_t *data /*[64]*/)
+              * On little-endian machines, we can process properly aligned
+              * data without copying it.
+              */
+-            if (!((data - (const md5_byte_t *)0) & 3)) {
++            if (!((reinterpret_cast<uintptr_t>(data)) & 3)) {
```
Comment (strega-nil-ms) [inline ports/zeroc-ice/md5i_fix.patch:10]: I do not like this being changed, why was it necessary?

File: ports/zeroc-ice/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    COMMAND powershell ${CURRENT_PORT_DIR}/fixProjFiles.ps1 "${SOURCE_PATH}"
+    LOGNAME fixProjFiles-${TARGET_TRIPLET}
+    WORKING_DIRECTORY ${CURRENT_BUILDTREES_DIR}
+  )
+  
+
+
+
+set(MSVC_TOOLSET_VER ${VCPKG_PLATFORM_TOOLSET})
+if(${VCPKG_PLATFORM_TOOLSET} STREQUAL "v144" OR ${VCPKG_PLATFORM_TOOLSET} STREQUAL "v145")
```
Comment (strega-nil-ms) [inline ports/zeroc-ice/portfile.cmake:161]: AFAIK, `v144` and `v145` are not out yet and won't be for three years (if they ever do come out)

File: ports/zeroc-ice/fixProjFiles.ps1 [hunk 6]
```diff
@@ -0,0 +1,41 @@
+
```
Comment (strega-nil-ms) [inline ports/zeroc-ice/fixProjFiles.ps1:1]: Can this be a CMake script instead?

File: ports/zeroc-ice/portfile.cmake [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+  # Setting these as environment variables, as .d files aren't generated the first time passing them to make.
+  set(ENV{MCPP_HOME} ${CURRENT_INSTALLED_DIR})
+  set(ENV{EXPAT_HOME} ${CURRENT_INSTALLED_DIR})
+  set(ENV{BZ2_HOME} ${CURRENT_INSTALLED_DIR})
+  set(ENV{LMDB_HOME} ${CURRENT_INSTALLED_DIR})
+  set(ENV{CPPFLAGS} "-I${CURRENT_INSTALLED_DIR}/include")
+  set(ENV{LDFLAGS} "-L${CURRENT_INSTALLED_DIR}/debug/lib")
+
+  set(ICE_BUILD_CONFIG "cpp11-shared")
+  if(${VCPKG_LIBRARY_LINKAGE} STREQUAL "static")
```
Comment (strega-nil-ms) [inline ports/zeroc-ice/portfile.cmake:112]:
```suggestion
  if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
```

File: ports/zeroc-ice/portfile.cmake [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    set(ICE_BUILD_CONFIG "cpp11-static")
+  endif()
+
+  message(STATUS "Building ${TARGET_TRIPLET}-dbg")
+  vcpkg_execute_build_process(
+    COMMAND make install V=1 prefix=${CURRENT_PACKAGES_DIR}/debug linux_id=vcpkg CONFIGS=${ICE_BUILD_CONFIG} USR_DIR_INSTALL=yes OPTIMIZE=no ${ICE_OPTIONAL_COMPONENTS_MAKE} -j${VCPKG_CONCURRENCY}
```
Comment (strega-nil-ms) [inline ports/zeroc-ice/portfile.cmake:118]:
```suggestion
    COMMAND make install
        V=1
        "prefix=${CURRENT_PACKAGES_DIR}/debug"
        linux_id=vcpkg
        "CONFIGS=${ICE_BUILD_CONFIG}"
        USR_DIR_INSTALL=yes
        OPTIMIZE=no
        ${ICE_OPTIONAL_COMPONENTS_MAKE}
        "-j${VCPKG_CONCURRENCY}"
```

**Review by strega-nil-ms — CHANGES_REQUESTED** [review]
Also, please format the CMake documents with 4 spaces.

File: scripts/ci.baseline.txt (hunk 2 above)
Comment (bold84) [reply scripts/ci.baseline.txt:1449]: Not yet. The build succeeds but the choice of components/features doesn't have any effect. I will work on this today, so the behavior on Windows and Linux/Mac will be the same.

File: ports/zeroc-ice/slice2swift.patch (hunk 3 above)
Comment (bold84) [reply ports/zeroc-ice/slice2swift.patch:22]: Clang 13 complains about typeCtx not being used. After looking at the code, and considering that we don't need the slice2swift compiler here anyways, I came to the conclusion that it's safe to remove it. 

When I'm done with the Linux/Mac part, this patch can be removed, as this project won't be built anyways.

File: ports/zeroc-ice/md5i_fix.patch (hunk 4 above)
Comment (bold84) [reply ports/zeroc-ice/md5i_fix.patch:10]: Clang 13: performing pointer subtraction with a null pointer may have undefined behavior

Do you have another solution? :-)

File: ports/zeroc-ice/portfile.cmake (hunk 5 above)
Comment (bold84) [reply ports/zeroc-ice/portfile.cmake:161]: Dunno, I just thought that if they come out and the zeroc-ice version hasn't been updated (no vs-version-specific solution file), there might be a good chance that this will prevent the build from failing by that time. 

I can remove it. Let me know what you think.

File: ports/zeroc-ice/fixProjFiles.ps1 (hunk 6 above)
Comment (bold84) [reply ports/zeroc-ice/fixProjFiles.ps1:1]: I thought about that too. 😃
Can I use the vcpkg downloads folder to cache the downloaded dlls for subsequent builds?

File: ports/zeroc-ice/md5i_fix.patch (hunk 4 above)
Comment (bold84) [reply ports/zeroc-ice/md5i_fix.patch:10]: https://github.com/zeroc-ice/ice/pull/1353

They don't seem to have an issue with it.

File: ports/zeroc-ice/vcpkg.json (hunk 1 above)
Comment (bold84) [reply ports/zeroc-ice/vcpkg.json:6]: The documentation advises to set "license" to null. The Ice has dual licensing. Specifying only GPL would be incorrect. And since the documentation refers to Qt, I decided to set it to null.

Screen Shot 2022-03-29 at 13 44 19

Source: https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/manifest-files.md#license

File: ports/zeroc-ice/slice2swift.patch (hunk 3 above)
Comment (bold84) [reply ports/zeroc-ice/slice2swift.patch:22]: https://github.com/zeroc-ice/ice/pull/1354

In the meanwhile I created a PR with this change here:

File: scripts/ci.baseline.txt (hunk 2 above)
Comment (bold84) [reply scripts/ci.baseline.txt]: Linux and MacOS builds should work now. 
At the moment static builds on the aforementioned platforms do not build executables like glacierrouter, icebox and so on. I have to manipulate the Makefile. I will take care of in the coming days.

*-windows-static unchaged, but *-windows-static-md works, as mentioned.

**Comment by bold84** [comment]
Done.

I apologize for abusing the PR by pushing to this branch many times. I used this branch to transfer changes between different operating systems and architectures. In the future I would use a separate branch for this purpose. 😁

The port should build for Linux(x86,x64,arm,arm64), MacOS (x86,x64,arm64) and Windows (x86,x64), in shared and static build configurations. ZeroC's VS project files do not have configurations for arm and it would need a bit of work to make this work, because the slice compiler has to be built for the host machine as it's needed during the build of the subsequent components.

I will add Android, iOS and possibly Webassembly in separate PRs in the near future. The latter needs some work on Ice itself though.

**Comment by bold84** [comment]
Why does it take so long for the commits to be build? Is there some kind of throttling?

**Comment by bold84** [comment]
> > Why does it take so long for the commits to be built? Is there some kind of throttling?
> 
> No, it is congestion.
> 
> * Recent baseline regressions caused a backlog.
> * A broken tool/CI interaction causes PR CI runs to build ports which are known to fail.
> * The recent zlib update invalidated a lot of cached artifacts.
> * Some of the queued PRs need a lot of build time.
> * At least one port with significant build time fails to upload to cache.
> 
> Take a look at the queue: https://dev.azure.com/vcpkg/public/_build?definitionId=27&_a=summary

Thank you for the explanation. 😊

**Comment by LilyWangLL** [comment]
@strega-nil-ms Could you please review this PR again?

**Comment by bold84** [comment]
How many weeks does it typically take until you guys review change requests? ☺️

**Comment by LilyWangLL** [comment]
Ping @strega-nil

**Review by strega-nil-ms — APPROVED** [review]
merging once green!

**Comment by ZeroAbility** [comment]
> Done.
> 
> I apologize for abusing the PR by pushing to this branch many times. I used this branch to transfer changes between different operating systems and architectures. In the future I would use a separate branch for this purpose. grin
> 
> The port should build for Linux(x86,x64,arm,arm64), MacOS (x86,x64,arm64) and Windows (x86,x64), in shared and static build configurations. ZeroC's VS project files do not have configurations for arm and it would need a bit of work to make this work, because the slice compiler has to be built for the host machine as it's needed during the build of the subsequent components.
> 
> I will add Android, iOS and possibly Webassembly in separate PRs in the near future. The latter needs some work on Ice itself though.

Perhaps these will help with icebuilder and arm:
https://github.com/mumble-voip/ice/commit/a9d68a8d1d884b839f871c46b4ea94b3fc6d6527
https://github.com/mumble-voip/ice/blob/3.7/icebuilder/CMakeLists.txt

**Comment by bold84** [comment]
@ZeroAbility Thanks! 

At the moment re-doing the skia port. I will look into this later.
